### PR TITLE
skip iptables sync if no endpoint changes

### LINKED ISSUE #26637: kube-proxy should ignore "spam" updates to endpoints
#### Environment
- 3 node Kubernetes v1.2.4 cluster with HA
- scheduler and controller-manager have `--leader-elect=true`
- no additional POD running aside from those in the `kube-system` namespace
#### Observed behavior
- `kube-proxy` runs `iptables-save/restore` twice per second.

Increasing `kube-proxy` verbosity shows that this is triggered by `OnEndpointsUpdate`

Stopping all the schedulers and controller-managers, then restarting only one instance of each without `--leader-elect` stops `kube-proxy` from running `iptables` that often.

### PR DESCRIPTION
**Pull request by freehan**
Alternative to https://github.com/kubernetes/kubernetes/pull/41173
fixes: #26637
No need to checksum. Just compare endpoint maps.

### CODE REVIEW

**Comment by freehan** [comment]
I think this one is cleaner than the checksum one. 
I did my best to sort out the original logic and build on top of it. It probably need a major refactor at some point. But since this PR is aiming for cherrypick, so only changing the minimum

File: pkg/proxy/iptables/proxier.go [hunk 1]
```diff
@@ -627,32 +628,38 @@ func (proxier *Proxier) OnEndpointsUpdate(allEndpoints []api.Endpoints) {
...
-				// Once the set operations using the list of ips are complete, build the list of endpoint infos
-				proxier.endpointsMap[svcPort] = proxier.buildEndpointInfoList(portsToEndpoints[portname], newEndpoints)
 			}
+			// Once the set operations using the list of ips are complete, build the list of endpoint infos
+			newEndpointsMap[svcPort] = proxier.buildEndpointInfoList(portsToEndpoints[portname], newEndpoints)
 			activeEndpoints[svcPort] = true
 		}
 	}
-	// Remove endpoints missing from the update.
+	// Gather stale connections to endpoints missing from the update.
```
Comment (bprashanth) [inline pkg/proxy/iptables/proxier.go]: nit: suggest rewording to "Check stale connections against endpoints missing from the update"

File: pkg/proxy/iptables/proxier.go [hunk 2]
```diff
@@ -627,32 +628,38 @@ func (proxier *Proxier) OnEndpointsUpdate(allEndpoints []api.Endpoints) {
...
 			if len(curEndpointIPs) != len(newEndpoints) || !slicesEquiv(slice.CopyStrings(curEndpointIPs), newEndpoints) {
+				glog.V(3).Infof("Setting endpoints for %q to %+v", svcPort, newEndpoints)
+				// Gather stale connections to removed endpoints
 				removedEndpoints := getRemovedEndpoints(curEndpointIPs, newEndpoints)
 				for _, ep := range removedEndpoints {
 					staleConnections[endpointServicePair{endpoint: ep, servicePortName: svcPort}] = true
 				}
-				glog.V(3).Infof("Setting endpoints for %q to %+v", svcPort, newEndpoints)
-				// Once the set operations using the list of ips are complete, build the list of endpoint infos
-				proxier.endpointsMap[svcPort] = proxier.buildEndpointInfoList(portsToEndpoints[portname], newEndpoints)
```
Comment (bprashanth) [inline pkg/proxy/iptables/proxier.go]: Comment: The only significant diff is that we were only doing the work to `buildEndpointInfoList` lazily, but that function only seems as expensive as `flattenEndpointsInfo` anyway (O(endpoints)), so I don't think we're burning any more cpu.

File: pkg/proxy/iptables/proxier.go [hunk 3]
```diff
@@ -627,32 +628,38 @@ func (proxier *Proxier) OnEndpointsUpdate(allEndpoints []api.Endpoints) {
...
+	// Gather stale connections to endpoints missing from the update.
 	for svcPort := range proxier.endpointsMap {
 		if !activeEndpoints[svcPort] {
+			glog.V(2).Infof("Removing endpoints for %q", svcPort)
 			// record endpoints of unactive service to stale connections
 			for _, ep := range proxier.endpointsMap[svcPort] {
 				staleConnections[endpointServicePair{endpoint: ep.ip, servicePortName: svcPort}] = true
 			}
-
-			glog.V(2).Infof("Removing endpoints for %q", svcPort)
```
Comment (bprashanth) [inline pkg/proxy/iptables/proxier.go]: Comment: this version seems cleaner since we're not deleting from the map we're iterating, and there's no residual state as we start with a fresh map everytime.

**Review by bprashanth — COMMENTED** [review]
/lgtm

**Comment by freehan** [comment]
Should we get this in and let it soak for some time before cherry-pick into 1.5?

**Comment by freehan** [comment]
adding label based on https://github.com/kubernetes/kubernetes/pull/41223#pullrequestreview-21161664

**Comment by dcbw** [comment]
@freehan I've had https://github.com/kubernetes/kubernetes/pull/39484 sitting around since forever that @thockin hasn't gotten to.  That adds a boatload of testcases.

And one of those testcases catches an error in this PR, which is that healthcheck updates are not correctly added anymore because proxier.updateHealthCheckEntries() is no longer called for every svcPort in the new endpoints map too.

I'm happy to fix that up in https://github.com/kubernetes/kubernetes/pull/39484 if you like, provided I can actually get somebody (@thockin ? @bprashanth ?) to review it.

**Comment by freehan** [comment]
@dcbw Thanks for noticing! 

Please go ahead with the fix in #39484.

**Comment by dcbw** [comment]
@freehan I've rebased and repushed #39484, would you mind taking a look at it too?  Thanks!

**Comment by thockin** [comment]
Thanks guys,

I have been out for about 2 weeks (combo of vaca, work, sick) and am now trying to get all these PRs accounted for.  Minhan can own the review.

I have another set of PRs between me, @timothysc, and @danwinship which make sync be called at a max rate - all together these should be a nice improvement.

Should we do this same transform over Service updates?

**Comment by dcbw** [comment]
@thockin yeah, we probably should do the same for Service updates, but it's not as big of an issue since Service updates are much less frequent than Endpoints.

**Comment by freehan** [comment]
@dcbw 
I understand the bug you are referring in https://github.com/kubernetes/kubernetes/pull/41223#issuecomment-279120970

I think that should not cause any problem at this point. Here is the theory:
1. create new service `svc1` 
2. OnEndpointsUpdate gets called with new service endpoint. 
3. Since `updateHealthCheckEntries` is only called for previous known service ports in proxier.endpointsMap, healthcheck for the newly added service endpoint will not be included.
4. OnEndpointsUpdate gets called again with the same of endpoints.
5. This time,  proxier.endpointsMap includes `svc1` and healthcheck is updated. 

Since OnEndpointsUpdate is called at least twice a second (Due to master election logic in kube-controller-manager and kube-scheduler), having a half second delay to update health check should not cause any problem, right? 


I intended to cherry-pick this into 1.5 to help some customers. I can add a surgical fix to combine svcPorts from newEndpointMap and proxier.endpointsMap, and trigger updateHealthCheckEntries. #39484 just need to rebased and delete the surgical fix. 

What do you think?
